### PR TITLE
HB_008: Update user information (firstName and lastName)

### PR DESCRIPTION
**Pull request by toanlc01**

### CODE REVIEW

File: server/src/User/user.controller.ts [hunk 1]
```diff
@@ -1,51 +1,37 @@
...
-import { Controller, Get, Param } from '@nestjs/common';
-import { LocalAuthGuard } from 'src/Auth/local.auth.guard';
-import { UserLoginDto } from 'src/Dto/user.dto';
+import { Request, UseGuards, HttpException, HttpStatus } from '@nestjs/common';
+import { Controller, Get, Param, Post, Body } from '@nestjs/common';
+import { JwtAuthGuard } from '../Auth/jwt-auth.guard';
 import { User } from './user.entity';
 import { UserService } from './user.service';
 
+// @UseGuards(JwtAuthGuard)
```
Comment (nguyen315) [inline server/src/User/user.controller.ts:7]: please remove this comment, useless comment

Comment (toanlc01) [reply server/src/User/user.controller.ts:7]: fixed

File: server/src/User/user.controller.ts [hunk 2]
```diff
@@ -1,51 +1,37 @@
...
   }
 
   @Post()
   addUser(@Body() user: User): Promise<User> {
     return this.userService.createUser(user);
   }
+
+  @Post(':id/update-user')
+  async updateUser(@Param('id') id: number, 
+             @Body('username') username: string,
```
Comment (nguyen315) [inline server/src/User/user.controller.ts:28]: We don't allow changing username, remove this field from post request body.

File: server/src/User/user.controller.ts [hunk 3]
```diff
@@ -1,51 +1,37 @@
...
+             @Body('firstname') firstname: string,
+             @Body('lastname') lastname: string) {
+    await this.userService.updateUsername(id, username)
```
Comment (nguyen315) [inline server/src/User/user.controller.ts:31]: Same as above

File: server/src/User/user.service.ts [hunk 4]
```diff
@@ -18,19 +20,81 @@ export class UserService {
...
-    return this.userRepository.save(newUser);
+    await this.userRepository.save(newUser);
+    return newUser;
   }
 
   findByUsername(userName: string): Promise<User | undefined> {
     return this.userRepository.findOne({ username: userName });
   }
+
+  async updateUsername(id: number, username: string) {
```
Comment (nguyen315) [inline server/src/User/user.service.ts:40]: Same as above

**Comment by toanlc01** [comment]
Fixed

File: server/src/User/user.service.ts [hunk 5]
```diff
@@ -18,19 +20,54 @@ export class UserService {
...
+    await this.userRepository.save(newUser);
+    return newUser;
   }
 
   findByUsername(userName: string): Promise<User | undefined> {
     return this.userRepository.findOne({ username: userName });
   }
+
+  updateLastName(id: number, lastname: string) {
+    if (this.checkFirstNameLength(lastname)) {
```
Comment (nguyen315) [inline server/src/User/user.service.ts:41]: Change name of function checkFirstNameLength to use in more contexts.

File: server/src/User/user.service.ts [hunk 6]
```diff
@@ -18,19 +20,54 @@ export class UserService {
...
+      throw new BadRequestException("last name length must be between 1 character and 20 characters")
+    }
+    return this.userRepository.update({id: id}, {lastname: lastname})
+  }
+
+  checkLastNameLength(lastName: string) {
```
Comment (nguyen315) [inline server/src/User/user.service.ts:47]: checkLastNameLength and checkFirstNameLength functions are same purpose

File: server/src/User/user.entity.ts [hunk 7]
```diff
@@ -22,6 +22,12 @@ export class User {
   @Column()
   password: string;
 
+  @Column()
+  firstname: string;
```
Comment (nguyen315) [inline server/src/User/user.entity.ts:26]: This should be firstName

File: server/src/User/user.entity.ts [hunk 8]
```diff
@@ -22,6 +22,12 @@ export class User {
...
+
+  @Column()
+  lastname: string;
```
Comment (nguyen315) [inline server/src/User/user.entity.ts:29]: And this should be lastName

**Comment by nguyen315** [comment]
Need to apply validate input for change password & update profile

**Pull request merged**